secret_store/src/key_server_cluster/cluster.rs

```diff
@@ -548,51 +533,98 @@ impl ClusterCore {
-					let session_state = session.state();
-					if session_state == DecryptionSessionState::Finished {
+					if session.is_finished() {
 						info!(target: "secretstore_net", "{}: decryption session completed", data.self_key_pair.public());
+						data.sessions.decryption_sessions.remove(&decryption_session_id);
+						break;
+					}
+
+					// try to dequeue message
+					match data.sessions.decryption_sessions.dequeue_message(&decryption_session_id) {
+						Some((msg_sender, msg)) => {
+							sender = msg_sender;
+							message = msg;
+						},
+						None => break,
+					}
+				},
+				Err(err) => {
+					warn!(target: "secretstore_net", "{}: decryption session error {} when processing message {} from node {}", data.self_key_pair.public(), err, message, sender);
+					data.sessions.respond_with_decryption_error(&session_id, &sub_session_id, &sender, message::DecryptionSessionError {
+						session: session_id.clone().into(),
+						sub_session: sub_session_id.clone().into(),
+						error: format!("{:?}", err),
+					});
+					if err != Error::InvalidSessionId {
+						data.sessions.decryption_sessions.remove(&decryption_session_id);
 					}
-					if session_state == DecryptionSessionState::Finished || session_state == DecryptionSessionState::Failed {
-						data.sessions.remove_decryption_session(&session_id, &sub_session_id);
+					break;
+				},
+			}
+		}
+	}
+
+	/// Process singlesigning message from the connection.
+	fn process_signing_message(data: Arc<ClusterData>, connection: Arc<Connection>, mut message: SigningMessage) {
+		let session_id = message.session_id().clone();
+		let sub_session_id = message.sub_session_id().clone();
+		let signing_session_id = SigningSessionId::new(session_id.clone(), sub_session_id.clone());
+		let mut sender = connection.node_id().clone();
+		let session = match message {
+			SigningMessage::SigningConsensusMessage(ref message) if match message.message {
+				ConsensusMessage::InitializeConsensusSession(_) => true,
+				_ => false,
+			} => {
+				let mut connected_nodes = data.connections.connected_nodes();
+				connected_nodes.insert(data.self_key_pair.public().clone());
+
+				let cluster = Arc::new(ClusterView::new(data.clone(), connected_nodes));
+				data.sessions.new_signing_session(sender.clone(), session_id.clone(), sub_session_id.clone(), cluster, None)
+			},
+			_ => {
+				data.sessions.signing_sessions.get(&signing_session_id)
+					.ok_or(Error::InvalidSessionId)
+			},
+		};
+
+		let mut is_queued_message = false;
+		loop {
+			match session.clone().and_then(|session| session.process_message(&sender, &message)) {
+				Ok(_) => {
+					// if session is completed => stop
+					let session = session.clone().expect("session.method() call finished with success; session exists; qed");
+					if session.is_finished() {
+						info!(target: "secretstore_net", "{}: signing session completed", data.self_key_pair.public());
```

Comment:
It is convenient for *me* right now :) In production environment it could be a little annoying, as long as SecretStore is run along with synchronization. But one day (I hope :)) it will became a standalone IPC-module && creating/completion sessions would be its main 'side effect' => it would be great to see these notifications in console.
I could remove this for now, if you think it is annoying.